### PR TITLE
[kotlin-server][Java][JAX-RS] Fix path shadowing (#23414)

### LINKED ISSUE #23414: [BUG][kotlin-server] Common path prefix extraction breaks cross-tag JAX-RS routing
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?

##### Description

In v7.21.0 the `kotlin-server` generator (with `jaxrs-spec` library) started extracting a common path prefix from each API tag's operations and placing it in the class-level `@Path` annotation. This breaks JAX-RS routing when routes from **different** tags share a path prefix.

Given this spec:

```yaml
paths:
  /foo/bar/one:
    get:
      tags: [tag-one]
      operationId: getOne
      # ...
  /foo/bar/two:
    get:
      tags: [tag-one]
      operationId: getTwo
      # ...
  /foo/bar/three:
    get:
      tags: [tag-two]
      operationId: getThree
      # ...
  /baz/bar/four:
    get:
      tags: [tag-two]
      operationId: getFour
      # ...
```

**v7.20.0** generates (correct):

```kotlin
// TagOneApi.kt
@Path("/")
interface TagOneApi {
    @GET @Path("/foo/bar/one") fun getOne(): Response
    @GET @Path("/foo/bar/two") fun getTwo(): Response
}

// TagTwoApi.kt
@Path("/")
interface TagTwoApi {
    @GET @Path("/baz/bar/four") fun getFour(): Response
    @GET @Path("/foo/bar/three") fun getThree(): Response
}
```

**v7.21.0** generates (broken):

```kotlin
// TagOneApi.kt
@Path("/foo/bar")   // <-- extracted common prefix
interface TagOneApi {
    @GET @Path("/one") fun getOne(): Response
    @GET @Path("/two") fun getTwo(): Response
}

// TagTwoApi.kt
@Path("")   // <-- no common prefix, but "/" changed to ""
interface TagTwoApi {
    @GET @Path("/baz/bar/four") fun getFour(): Response
    @GET @Path("/foo/bar/three") fun getThree(): Response
}
```

`TagOneApi` claims `@Path("/foo/bar")` at the class level. The JAX-RS runtime (e.g. Quarkus REST) routes `GET /foo/bar/three` to `TagOneApi` since it has the most specific class-level prefix match. `TagOneApi` has no `/three` method, so the request returns **404**. `TagTwoApi`'s route is never reached.

##### openapi-generator version

7.21.0 (regression from 7.20.0)

##### OpenAPI declaration file content or url

Find a full reproducer at https://github.com/lbilger/openapi-generator-path-shadowing-reproducer.

##### Generation Details

See linked reproducer.

##### Steps to reproduce

- Clone the [reproducer](https://github.com/lbilger/openapi-generator-path-shadowing-reproducer)
- Tests pass with 7.20.0:

  ```sh
  mvn clean test -Pv7.20.0
  ```

- Tests fail with 7.21.0:

  ```sh
  mvn clean test -Pv7.21.0
  ```

##### Related issues/PRs

None found.

##### Suggest a fix

Revert the common path extraction code or make it optional by introducing a configuration property.

Alternatively, detect if path prefixes are used in another tag and prevent extraction then.

##### Note

I have used AI to help me create the reproducer and the text for this ticket, but have verified and edited them manually.

### PR DESCRIPTION
**Pull request by feczkob**
Fixes https://github.com/OpenAPITools/openapi-generator/issues/23414.

The generator now tracks all resource paths across tags during code generation. Before applying a common `@Path` prefix to an API class, it checks if that prefix would "shadow" (intercept) routes belonging to other tags - if so, it keeps the full paths on each method instead. 

Could you take a look, please?

@karismann (2019/03) @Zomzog (2019/04) @andrewemery (2019/10) @4brunu (2019/11) @yutaka0m (2020/03) @stefankoppier (2022/06) @e5l (2024/10) @dennisameling (2026/02) and @lbilger 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package || exit
  ./bin/generate-samples.sh ./bin/configs/*.yaml || exit
  ./bin/utils/export_docs_generators.sh || exit
  ``` 
  (For Windows users, please run the script in [WSL](https://learn.microsoft.com/en-us/windows/wsl/install))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR solves a reported issue, reference it using [GitHub's linking syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) (e.g., having `"fixes #123"` present in the PR description)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

---
## Summary by cubic
Fixes path shadowing in Java JAX-RS and Kotlin `jaxrs-spec` generators by skipping a class-level `@Path` when it would hide routes from other tags, keeping all endpoints reachable. Fixes #23414.

- **Bug Fixes**
  - Collect all resource paths across tags and skip `commonPath` if it would shadow another tag’s route.
  - Add a cross-tag shadow check in `AbstractJavaJAXRSServerCodegen` during `jaxrsPostProcessOperations`; include helpers to compute/apply common path.
  - Mirror the logic in `KotlinServerCodegen` for the `jaxrs-spec` library; other Kotlin libraries remain unchanged.
  - Add regression tests and sample spec: `JavaJaxrsBaseTest#testCommonPathDoesNotShadowOtherTags`, `KotlinServerCodegenTest#testCommonPathDoesNotShadowOtherTags_jaxrsSpec`, `src/test/resources/3_0/issue_23414.yaml`.
  - Regenerate samples to reflect the new path handling (e.g., JAX-RS `FakeApi` now uses `@Path("")` with full method paths; add Kotlin Ktor sample files `AppMain.kt` and `Configuration.kt`).

<sup>Written for commit 0986dc716c176548042b1d7c9234eca8c97d1706. Summary will update on new commits. <a href="https://cubic.dev/pr/OpenAPITools/openapi-generator/pull/23871?utm_source=github">Review in cubic</a></sup>

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaJAXRSServerCodegen.java [hunk 1]
```diff
@@ -70,6 +81,15 @@ public abstract class AbstractJavaJAXRSServerCodegen extends AbstractJavaCodegen
...
     protected boolean useTags = false;
 
+    /**
+     * All resource paths seen across every tag, collected during the first pass
+     * ({@link #addOperationToGroup}).  Used in the second pass
+     * ({@link #postProcessOperationsWithModels}) to detect cross-tag path shadowing: a
+     * candidate {@code commonPath} is only safe if no <em>other</em> tag owns a resource
+     * path that starts with that prefix.
+     */
+    private final Set<String> allResourcePaths = new HashSet<>();
```
Comment (feczkob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaJAXRSServerCodegen.java:91]: Global path registry.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaJAXRSServerCodegen.java [hunk 2]
```diff
@@ -138,6 +158,8 @@ public void addOperationToGroup(String tag, String resourcePath, Operation opera
             final List<CodegenOperation> opList = operations.computeIfAbsent(co.baseName, k -> new ArrayList<>());
             opList.add(co);
         }
+
+        allResourcePaths.add(resourcePath);
```
Comment (feczkob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaJAXRSServerCodegen.java:162]: Every time an operation is assigned to a tag, its full resource path is added to the global path registry.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaJAXRSServerCodegen.java [hunk 3]
```diff
@@ -193,99 +215,164 @@ public OperationsMap postProcessOperationsWithModels(OperationsMap objs, List<Mo
...
+        processOperationMetadata(ops);
+
+        String commonPath = computeCommonPath(ops);
+
+        if (commonPath != null && !commonPath.isEmpty() && !"/".equals(commonPath)
+                && wouldShadowOtherTags(commonPath, ops, allResourcePaths)) {
+            commonPath = null;
+        }
+
+        applyCommonPath(ops, commonPath, objs);
```
Comment (feczkob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaJAXRSServerCodegen.java:240]: Before applying a common path prefix to a tag's API class, the generator checks shadowing.

**Comment by wing328** [comment]
@feczkob thanks for the PR

@lbilger I wonder if you can test/review this change to confirm whether it addresses the issue https://github.com/OpenAPITools/openapi-generator/issues/23414

```
git checkout -b feczkob-fix/23414-path-shadowing master
git pull https://github.com/feczkob/openapi-generator.git fix/23414-path-shadowing
```

**Comment by wing328** [comment]
@feczkob are you using this fix in production and confirm it works for your use cases?

**Comment by feczkob** [comment]
@wing328 no, the bug appears only in an edge case. However, I verified the change on the reproducer of @lbilger and the tests pass. I also took a look at the generated classes and they are as expected. Not sure whether we should wait more for his feedback or we can rather proceed with the merge.

**Comment by lbilger** [comment]
Hi @feczkob , thanks a lot for fixing this! Unfortunately, I can't test it at the moment, but it looks good and if you tried with the reproducer, I'm inclined to believe that it works. So thumbs up from me 👍

**Comment by feczkob** [comment]
@lbilger let's test it when you have some free time, just to be super sure 😄 In the meantime, @wing328 do you want to wait more, or can we proceed with the merge?

**Comment by wing328** [comment]
Let's give it a try 👍

**Comment by lbilger** [comment]
Hi @feczkob , I found some time to test it with my real project. I built 7.24.0-SNAPSHOT from `master` and it fixed the problem 🎉 

Thanks a lot for the fix!
